internal/build/imgsrc/tokens.go

```diff
@@ -0,0 +1,211 @@
+	cfg := config.FromContext(ctx)
+
+	// If we have an unexpired token for this organization, return it
+	if cachedToken, ok := cfg.CachedBuildTokens[orgID]; ok {
+		expired := time.Now().Add(time.Minute).After(cachedToken.Expiration)
+		if !expired {
+			return cachedToken.Token, nil
+		}
+	}
+
+	// Otherwise, we need to create a token for this organization
+	apiClient := fly.ClientFromContext(ctx)
+	resp, err := gql.CreateLimitedAccessToken(
+		ctx,
+		apiClient.GenqClient,
+		buildTokenName,
+		org.ID,
+		"deploy_organization",
+		&gql.LimitedAccessTokenOptions{},
+		buildTokenExpiry.String(),
+	)
+	if err != nil {
+		return "", err
+	}
+
+	toks, err := macaroon.Parse(resp.CreateLimitedAccessToken.LimitedAccessToken.TokenHeader)
+	if err != nil {
+		return "", err
+	}
+
+	perms, _, disMacs, disToks, err := macaroon.FindPermissionAndDischargeTokens(toks, flyio.LocationPermission)
+	if err != nil {
+		return "", err
+	}
+	if len(perms) != 1 {
+		return "", errors.New("expected exactly one permission token")
+	}
+
+	// Mask access, but skip expiry because we already specified it
+	m := perms[0]
+	addBuildTokenCaveats(m, orgID, false)
+
+	perm, err := m.Encode()
+	if err != nil {
+		return "", err
+	}
+
+	token := encodeMacaroons(append([][]byte{perm}, disToks...))
+
+	// Find the earliest time any of the tokens expire
+	expiration := m.Expiration()
+	for _, dis := range disMacs {
+		if e := dis.Expiration(); e.Before(expiration) {
+			expiration = e
+		}
+	}
+
+	// Cache token and expiration time
+	if cfg.CachedBuildTokens == nil {
+		cfg.CachedBuildTokens = make(map[uint64]config.CachedBuildToken)
+	}
+	cfg.CachedBuildTokens[orgID] = config.CachedBuildToken{
+		ID:         resp.CreateLimitedAccessToken.LimitedAccessToken.Id,
+		Token:      token,
+		Expiration: expiration,
+	}
+
+	return token, nil
+}
+
+func getBuildTokenFromMacaroons(orgID uint64, macaroons string) (string, error) {
+	raws, err := macaroon.Parse(macaroons)
+	if err != nil {
+		return "", err
+	}
+
+	perms, _, disMacs, disToks, err := macaroon.FindPermissionAndDischargeTokens(raws, flyio.LocationPermission)
+	if err != nil {
+		return "", err
+	}
+
+	dischargeByTicket := make(map[string][]byte)
+	for i, m := range disMacs {
+		dischargeByTicket[hex.EncodeToString(m.Nonce.KID)] = disToks[i]
+	}
+
+	var toks [][]byte
+	for _, m := range perms {
+		// Skip tokens for other organizations
+		orgScope, err := flyio.OrganizationScope(&m.UnsafeCaveats)
+		if err != nil || orgScope != orgID {
+			continue
+		}
+
+		// Mask access and add expiry
+		addBuildTokenCaveats(m, orgID, true)
+
+		tok, err := m.Encode()
+		if err != nil {
+			return "", fmt.Errorf("unable to encode macaroon: %w", err)
+		}
+
+		toks = append(toks, tok)
+
+		// Append all the relevant discharge tokens
+		for _, cav := range macaroon.GetCaveats[*macaroon.Caveat3P](&m.UnsafeCaveats) {
+			toks = append(toks, dischargeByTicket[hex.EncodeToString(cav.Ticket)])
```

Comment:
Should we return an error here if we don't have the discharge?

Reply:
I don't think we want to return an error, just skip this permission token entirely so we don't fail if there was another one which we did have all the discharges for. (If there weren't, we'd hit the `len(toks) == 0` below).

I was putting off refactoring that to whenever I rip out the discharge collection code into a `macaroon` helper function. That being said, it should probably be wrapped in `if dis, ok := dischargeByTicket[hex.EncodeToString(cav.Ticket)]; ok`.

Reply:
Makes sense. Right now, we're adding the permission token even if we don't have the discharges for it though, right?

Reply:
That's right. Does that cause issues, apart from being a bit wasteful?

Reply:
I don't think it's harmful